contrib/pg_tde/src/access/pg_tde_tdemap.c

```diff
@@ -1431,6 +1535,178 @@ pg_tde_get_key_from_cache(const RelFileLocator *rlocator, uint32 key_type)
 	return NULL;
 }
 
+/* 
+ * Returns last (the most recent) key for a given relation
+ */
+WALKeyCacheRec *
+pg_tde_get_last_wal_key(void)
+{
+	return tde_wal_key_last_rec;
+}
+
+WALKeyCacheRec *
+pg_tde_get_wal_cache_keys(void)
+{
+	return tde_wal_key_cache;
+}
+
+InternalKey *
+pg_tde_read_last_wal_key(void)
+{
+	RelFileLocator rlocator = GLOBAL_SPACE_RLOCATOR(XLOG_TDE_OID);
+	char		db_keydata_path[MAXPGPATH] = {0};
+	off_t		read_pos = 0;
+	LWLock	   *lock_pk = tde_lwlock_enc_keys();
+	TDEPrincipalKey *principal_key;
+	int			fd = -1;
+	int 		file_idx = 0;
+	bool		is_new;
+	InternalKey 	*enc_rel_key_data,
+					*rel_key_data;
+	off_t		fsize;
+
+	LWLockAcquire(lock_pk, LW_EXCLUSIVE);
+	principal_key = GetPrincipalKey(rlocator.dbOid, LW_EXCLUSIVE);
+	if (principal_key == NULL)
+	{
+		LWLockRelease(lock_pk);
+		elog(DEBUG1, "init WAL encryption: no principal key");
+		return NULL;
+	}
+	pg_tde_set_db_file_paths(rlocator.dbOid, NULL, db_keydata_path);
+
+	fd = pg_tde_open_file(db_keydata_path, &principal_key->keyInfo, false, O_RDONLY, &is_new, &read_pos);
```

Comment:
I'm not sure if this is also an issue elsewhere, but `pg_tde_open_file` can report an error, and we won't release the lock in that case.